File: ketoan/gen_hv_mapping.py
import json
from collections import defaultdict

with open('/tmp/hv_tonghop_items.json', 'r') as f:
    items = json.load(f)

# Sort by frequency
items.sort(key=lambda x: x['sl'], reverse=True)

def get_mapping(ten_goc):
    ten_goc_upper = ten_goc.upper()
    ma_hang = "-"
    ten_chuan = "-"
    
    # Same logic as before
    if any(keyword in ten_goc_upper for keyword in ["LAPTOP", "MÁY TÍNH XÁCH TAY", "MÁY VI TÌNH XÁCH TAY"]):
        if "DELL" in ten_goc_upper: return "LT-DELL", "Laptop Dell"
        elif "HP" in ten_goc_upper: return "LT-HP", "Laptop HP"
        elif "ASUS" in ten_goc_upper: return "LT-ASUS", "Laptop ASUS"
        elif "LENOVO" in ten_goc_upper: return "LT-LNV", "Laptop Lenovo"
        elif "MACBOOK" in ten_goc_upper or "APPLE" in ten_goc_upper: return "LT-MAC", "Apple MacBook"
        return "LT-AI", "Laptop văn phòng"
            
    elif "MÁY VI TÍNH" in ten_goc_upper or "BỘ MÁY" in ten_goc_upper or "DESKTOP" in ten_goc_upper:
        return "PC-AI", "Bộ máy vi tính để bàn"

    elif "MÁY IN" in ten_goc_upper:
        if "CANON" in ten_goc_upper: return "PR-CANON", "Máy in Canon"
        elif "BROTHER" in ten_goc_upper: return "PR-BRT", "Máy in Brother"
        elif "EPSON" in ten_goc_upper: return "PR-EPSON", "Máy in Epson"
        return "PR-AI", "Máy in"
            
    elif "MỰC" in ten_goc_upper or "CATRIDGE" in ten_goc_upper or "CARTRIDGE" in ten_goc_upper:
        if "CANON" in ten_goc_upper: return "INK-CANON", "Mực in Canon"
        elif "BROTHER" in ten_goc_upper: return "INK-BRT", "Mực in Brother"
        elif "EPSON" in ten_goc_upper: return "INK-EPSON", "Mực in Epson"
        return "INK-AI", "Mực in / Hộp mực"
            
    elif "SSD" in ten_goc_upper or "Ổ CỨNG" in ten_goc_upper or "HDD" in ten_goc_upper:
        return "SSD-AI", "Ổ cứng"
    elif "RAM" in ten_goc_upper or "BỘ NHỚ TRONG" in ten_goc_upper:
        return "RAM-AI", "Bộ nhớ RAM"
    elif "CHUỘT" in ten_goc_upper or "MOUSE" in ten_goc_upper:
        if "LOGITECH" in ten_goc_upper: return "MOUSE-LOGI", "Chuột máy tính Logitech"
        elif "RAPOO" in ten_goc_upper: return "MOUSE-RAPOO", "Chuột máy tính Rapoo"
        return "MOUSE-AI", "Chuột máy tính"
    elif "BÀN PHÍM" in ten_goc_upper or "KEYBOARD" in ten_goc_upper:
        return "KB-AI", "Bàn phím máy tính"
    elif "MÀN HÌNH" in ten_goc_upper or "MONITOR" in ten_goc_upper or "DISPLAY" in ten_goc_upper:
        if "DELL" in ten_goc_upper: return "MON-DELL", "Màn hình máy tính Dell"
        elif "SAMSUNG" in ten_goc_upper: return "MON-SAM", "Màn hình máy tính Samsung"
        elif "VIEWSONIC" in ten_goc_upper: return "MON-VIEW", "Màn hình máy tính ViewSonic"
        return "MON-AI", "Màn hình máy tính"
    elif "USB" in ten_goc_upper or "BỘ NHỚ NGOÀI" in ten_goc_upper or "THẺ NHỚ" in ten_goc_upper:
        return "USB-AI", "Bộ nhớ ngoài / USB"
        
    elif "CAMERA" in ten_goc_upper or "CCTV" in ten_goc_upper:
        return "CAM-AI", "Camera quan sát"
    elif "WEBCAM" in ten_goc_upper or "GHÌNH" in ten_goc_upper or "TRUYỀN HÌNH ẢNH" in ten_goc_upper:
        return "CAM-WCAM", "Webcam máy tính"
    elif "NGUỒN" in ten_goc_upper or "POWER" in ten_goc_upper:
        return "PSU-AI", "Nguồn máy tính"
    elif "MAINBOARD" in ten_goc_upper or "BO MẠCH CHỦ" in ten_goc_upper or "BẢNG MẠCH CHÍNH" in ten_goc_upper:
        return "MAIN-AI", "Bo mạch chủ"
    elif "CHÍP VI XỬ LÝ" in ten_goc_upper or "BỘ VI XỬ LÝ" in ten_goc_upper or "CPU" in ten_goc_upper or "PROCESSOR" in ten_goc_upper:
        return "CPU-AI", "Bộ vi xử lý (CPU)"
    elif "CÁP" in ten_goc_upper and ("DỮ LIỆU" in ten_goc_upper or "TÍN HIỆU" in ten_goc_upper or "HDMI" in ten_goc_upper):
        return "CBL-AI", "Cáp tín hiệu / Dữ liệu"
    elif "LOA" in ten_goc_upper or "SPEAKER" in ten_goc_upper:
        return "SPK-AI", "Loa vi tính"
    elif "TAI NGHE" in ten_goc_upper or "HEADPHONE" in ten_goc_upper:
        return "HP-AI", "Tai nghe vi tính"
    elif "PHẦN MỀM" in ten_goc_upper or "SOFTWARE" in ten_goc_upper or "KASPERSKY" in ten_goc_upper or "BẢN QUYỀN" in ten_goc_upper:
        return "SW-AI", "Bản quyền phần mềm"
    elif "BỘ LƯU ĐIỆN" in ten_goc_upper or "UPS" in ten_goc_upper:
        return "UPS-AI", "Bộ lưu điện (UPS)"
    elif "MÁY CHIẾU" in ten_goc_upper or "PROJECTOR" in ten_goc_upper:
        return "PRJ-AI", "Máy chiếu"
    elif "MÀN CHIẾU" in ten_goc_upper:
        return "SCR-PRJ", "Màn chiếu"
    elif "MÁY QUÉT" in ten_goc_upper or "MÁY ĐỌC MÃ" in ten_goc_upper or "MÁY CHẤM CÔNG" in ten_goc_upper or "MÁY IN HÓA ĐƠN" in ten_goc_upper or "SCANNER" in ten_goc_upper:
        return "OFFICE-EQ", "Thiết bị văn phòng (Quyét mã/Chấm công/In bills)"
        
    elif "ROUTER" in ten_goc_upper or "BỘ ĐỊNH TUYẾN" in ten_goc_upper or "THIẾT BỊ ĐỊNH TUYẾN" in ten_goc_upper:
        return "NW-ROUT", "Thiết bị định tuyến (Router)"
    elif "SWITCH" in ten_goc_upper or "CHUYỂN MẠCH" in ten_goc_upper:
        return "NW-SW", "Thiết bị chuyển mạch (Switch)"
    elif "WIFI" in ten_goc_upper or "THU PHÁT" in ten_goc_upper or "BỘ CHUYỂN ĐỔI" in ten_goc_upper:
        return "NW-WIFI", "Thiết bị Wifi/Chuyển đổi"
    elif "CÁP MẠNG" in ten_goc_upper:
        return "NW-CBL", "Cáp mạng"
        
    elif "BẢO TRÌ" in ten_goc_upper or "CÀI ĐẶT" in ten_goc_upper or "DỊCH VỤ" in ten_goc_upper or "THI CÔNG" in ten_goc_upper or "SỬA CHỮA" in ten_goc_upper or "THIẾT KẾ" in ten_goc_upper:
        return "SV-IT", "Dịch vụ IT / Bảo trì"
    elif "PHÍ" in ten_goc_upper or "CƯỚC" in ten_goc_upper or "DỊCH VỤ NGÂN HÀNG" in ten_goc_upper:
        return "FEE-AI", "Phí dịch vụ"
    elif "CHIẾT KHẤU" in ten_goc_upper or "GIẢM GIÁ" in ten_goc_upper:
        return "DISC-AI", "Chiết khấu thương mại"
    elif "KHUYẾN" in ten_goc_upper or "HÀNG TẶNG" in ten_goc_upper:
        return "GIFT-AI", "Hàng khuyến mãi"
    elif "LÃI" in ten_goc_upper and "VAY" in ten_goc_upper:
        return "FIN-AI", "Lãi ngân hàng/Lãi vay"
    elif "THANH TOÁN" in ten_goc_upper:
        return "PAY-AI", "Thanh toán giao dịch"
        
    else:
        return "IT-GEN", "Phụ kiện / Linh kiện chung"


groups = defaultdict(lambda: {"ma": "", "chuan": "", "sl_tong": 0, "gocs": set(), "goc_list": []})

for item in items:
    # Bỏ qua các mặt hàng quá dị (xuất hiện < 5 lần) để table gọn gàng hơn
    # Tuy nhiên vì group lại nên ta có thể giữ nguyên sl >= 2
    if item['sl'] < 2: continue
    
    ten_goc = item['tenGoc'].replace('\n', ' ').strip()
    ma_hang, ten_chuan = get_mapping(ten_goc)
    
    grp = groups[ma_hang]
    grp['ma'] = ma_hang
    grp['chuan'] = ten_chuan
    grp['sl_tong'] += item['sl']
    if ten_goc not in grp['gocs']:
        grp['gocs'].add(ten_goc)
        # Chỉ lưu max 5 tên gốc làm ví dụ minh họa
        if len(grp['goc_list']) < 5:
            grp['goc_list'].append(ten_goc)

# Generate Markdown
out = []
out.append(f"# Bảng Từ Điển Đồng Nghĩa (Synonyms) AI - CÔNG TY TNHH HUY VŨ")
out.append("")
out.append("Theo phân tích từ 1,968 mặt hàng phát sinh thực tế, AI đã gom nhóm và chuẩn hóa thành các danh mục sau. Bảng danh mục này được tối ưu dành riêng cho việc cấu hình RAG Synonyms:")
out.append("")
out.append("| STT | Mã Hàng (AI) | Tên Chuẩn Hóa | Tổng SL | Các Biến Thể Tên Hàng Thực Tế (Ví dụ) |")
out.append("| :--- | :--- | :--- | :--- | :--- |")

sorted_groups = sorted(groups.values(), key=lambda x: x['sl_tong'], reverse=True)

for idx, grp in enumerate(sorted_groups, 1):
    # Nối các biến thể bằng thẻ <br> để xuống dòng trong bảng
    # hoặc dùng dấu bullet
    variants = "<br>- ".join([""] + grp['goc_list'])
    # Nếu còn nữa thì thêm ...
    if len(grp['gocs']) > 5:
        variants += f"<br>*(...và {len(grp['gocs']) - 5} biến thể khác)*"
        
    out.append(f"| {idx} | **{grp['ma']}** | **{grp['chuan']}** | {grp['sl_tong']} | {variants[4:]} |")

out.append("")
out.append(f"*Ghi chú: Quá trình tinh gọn này đã giảm từ gần 2,000 danh mục thô xuống còn **{len(groups)} mã chuẩn hóa chính**, giúp RAG hoạt động cực kỳ chính xác và tiết kiệm Token LLM.*")

with open('/mnt/chikiet/kata2025/ragketoan/docs/huyvu/dulieuchuan/mapping_items.md', 'w') as f:
    f.write('\n'.join(out))

File: ketoan/gen_hv_mapping.js
const fs = require('fs');

const items = JSON.parse(fs.readFileSync('/tmp/hv_tonghop_items.json', 'utf8'));

// Sort by sl desc
items.sort((a, b) => b.sl - a.sl);

function getMapping(tenGoc) {
    const tenGocUpper = tenGoc.toUpperCase();
    
    // Grouping Logic
    if (tenGocUpper.includes("LAPTOP") || tenGocUpper.includes("MÁY TÍNH XÁCH TAY") || tenGocUpper.includes("MÁY VI TÌNH XÁCH TAY")) {
        if (tenGocUpper.includes("DELL")) return ["LT-DELL", "Laptop Dell"];
        if (tenGocUpper.includes("HP")) return ["LT-HP", "Laptop HP"];
        if (tenGocUpper.includes("ASUS")) return ["LT-ASUS", "Laptop ASUS"];
        if (tenGocUpper.includes("LENOVO")) return ["LT-LNV", "Laptop Lenovo"];
        if (tenGocUpper.includes("MACBOOK") || tenGocUpper.includes("APPLE")) return ["LT-MAC", "Apple MacBook"];
        return ["LT-AI", "Laptop văn phòng"];
    }
            
    if (tenGocUpper.includes("MÁY VI TÍNH") || tenGocUpper.includes("BỘ MÁY") || tenGocUpper.includes("DESKTOP")) {
        return ["PC-AI", "Bộ máy vi tính để bàn"];
    }

    if (tenGocUpper.includes("MÁY IN")) {
        if (tenGocUpper.includes("CANON")) return ["PR-CANON", "Máy in Canon"];
        if (tenGocUpper.includes("BROTHER")) return ["PR-BRT", "Máy in Brother"];
        if (tenGocUpper.includes("EPSON")) return ["PR-EPSON", "Máy in Epson"];
        return ["PR-AI", "Máy in"];
    }
            
    if (tenGocUpper.includes("MỰC") || tenGocUpper.includes("CATRIDGE") || tenGocUpper.includes("CARTRIDGE")) {
        if (tenGocUpper.includes("CANON")) return ["INK-CANON", "Mực in Canon"];
        if (tenGocUpper.includes("BROTHER")) return ["INK-BRT", "Mực in Brother"];
        if (tenGocUpper.includes("EPSON")) return ["INK-EPSON", "Mực in Epson"];
        return ["INK-AI", "Mực in / Hộp mực"];
    }
            
    if (tenGocUpper.includes("SSD") || tenGocUpper.includes("Ổ CỨNG") || tenGocUpper.includes("HDD")) return ["SSD-AI", "Ổ cứng"];
    if (tenGocUpper.includes("RAM") || tenGocUpper.includes("BỘ NHỚ TRONG")) return ["RAM-AI", "Bộ nhớ RAM"];
    
    if (tenGocUpper.includes("CHUỘT") || tenGocUpper.includes("MOUSE")) {
        if (tenGocUpper.includes("LOGITECH")) return ["MOUSE-LOGI", "Chuột máy tính Logitech"];
        if (tenGocUpper.includes("RAPOO")) return ["MOUSE-RAPOO", "Chuột máy tính Rapoo"];
        return ["MOUSE-AI", "Chuột máy tính"];
    }
    
    if (tenGocUpper.includes("BÀN PHÍM") || tenGocUpper.includes("KEYBOARD")) return ["KB-AI", "Bàn phím máy tính"];
    
    if (tenGocUpper.includes("MÀN HÌNH") || tenGocUpper.includes("MONITOR") || tenGocUpper.includes("DISPLAY")) {
        if (tenGocUpper.includes("DELL")) return ["MON-DELL", "Màn hình máy tính Dell"];
        if (tenGocUpper.includes("SAMSUNG")) return ["MON-SAM", "Màn hình máy tính Samsung"];
        if (tenGocUpper.includes("VIEWSONIC")) return ["MON-VIEW", "Màn hình máy tính ViewSonic"];
        return ["MON-AI", "Màn hình máy tính"];
    }
    
    if (tenGocUpper.includes("USB") || tenGocUpper.includes("BỘ NHỚ NGOÀI") || tenGocUpper.includes("THẺ NHỚ")) return ["USB-AI", "Bộ nhớ ngoài / USB"];
        
    if (tenGocUpper.includes("CAMERA") || tenGocUpper.includes("CCTV")) return ["CAM-AI", "Camera quan sát"];
    if (tenGocUpper.includes("WEBCAM") || tenGocUpper.includes("GHÌNH") || tenGocUpper.includes("TRUYỀN HÌNH ẢNH")) return ["CAM-WCAM", "Webcam máy tính"];
    if (tenGocUpper.includes("NGUỒN") || tenGocUpper.includes("POWER")) return ["PSU-AI", "Nguồn máy tính"];
    if (tenGocUpper.includes("MAINBOARD") || tenGocUpper.includes("BO MẠCH CHỦ") || tenGocUpper.includes("BẢNG MẠCH CHÍNH")) return ["MAIN-AI", "Bo mạch chủ"];
    if (tenGocUpper.includes("CHÍP VI XỬ LÝ") || tenGocUpper.includes("BỘ VI XỬ LÝ") || tenGocUpper.includes("CPU") || tenGocUpper.includes("PROCESSOR")) return ["CPU-AI", "Bộ vi xử lý (CPU)"];
    
    if (tenGocUpper.includes("CÁP") && (tenGocUpper.includes("DỮ LIỆU") || tenGocUpper.includes("TÍN HIỆU") || tenGocUpper.includes("HDMI"))) return ["CBL-AI", "Cáp tín hiệu / Dữ liệu"];
    
    if (tenGocUpper.includes("LOA") || tenGocUpper.includes("SPEAKER")) return ["SPK-AI", "Loa vi tính"];
    if (tenGocUpper.includes("TAI NGHE") || tenGocUpper.includes("HEADPHONE")) return ["HP-AI", "Tai nghe vi tính"];
    if (tenGocUpper.includes("PHẦN MỀM") || tenGocUpper.includes("SOFTWARE") || tenGocUpper.includes("KASPERSKY") || tenGocUpper.includes("BẢN QUYỀN")) return ["SW-AI", "Bản quyền phần mềm"];
    if (tenGocUpper.includes("BỘ LƯU ĐIỆN") || tenGocUpper.includes("UPS")) return ["UPS-AI", "Bộ lưu điện (UPS)"];
    if (tenGocUpper.includes("MÁY CHIẾU") || tenGocUpper.includes("PROJECTOR")) return ["PRJ-AI", "Máy chiếu"];
    if (tenGocUpper.includes("MÀN CHIẾU")) return ["SCR-PRJ", "Màn chiếu"];
    if (tenGocUpper.includes("MÁY QUÉT") || tenGocUpper.includes("MÁY ĐỌC MÃ VẠCH") || tenGocUpper.includes("MÁY CHẤM CÔNG") || tenGocUpper.includes("MÁY IN HÓA ĐƠN") || tenGocUpper.includes("SCANNER")) return ["OFFICE-EQ", "Thiết bị văn phòng (Quyét mã/Chấm công/In bills)"];
        
    if (tenGocUpper.includes("ROUTER") || tenGocUpper.includes("BỘ ĐỊNH TUYẾN") || tenGocUpper.includes("THIẾT BỊ ĐỊNH TUYẾN")) return ["NW-ROUT", "Thiết bị định tuyến (Router)"];
    if (tenGocUpper.includes("SWITCH") || tenGocUpper.includes("CHUYỂN MẠCH")) return ["NW-SW", "Thiết bị chuyển mạch (Switch)"];
    if (tenGocUpper.includes("WIFI") || tenGocUpper.includes("THU PHÁT") || tenGocUpper.includes("BỘ CHUYỂN ĐỔI")) return ["NW-WIFI", "Thiết bị Wifi/Chuyển đổi"];
    if (tenGocUpper.includes("CÁP MẠNG")) return ["NW-CBL", "Cáp mạng"];
        
    if (tenGocUpper.includes("BẢO TRÌ") || tenGocUpper.includes("CÀI ĐẶT") || tenGocUpper.includes("DỊCH VỤ") || tenGocUpper.includes("THI CÔNG") || tenGocUpper.includes("SỬA CHỮA") || tenGocUpper.includes("THIẾT KẾ")) return ["SV-IT", "Dịch vụ IT / Bảo trì"];
    if (tenGocUpper.includes("PHÍ") || tenGocUpper.includes("CƯỚC") || tenGocUpper.includes("DỊCH VỤ NGÂN HÀNG")) return ["FEE-AI", "Phí dịch vụ"];
    if (tenGocUpper.includes("CHIẾT KHẤU") || tenGocUpper.includes("GIẢM GIÁ")) return ["DISC-AI", "Chiết khấu thương mại"];
    if (tenGocUpper.includes("KHUYẾN") || tenGocUpper.includes("HÀNG TẶNG")) return ["GIFT-AI", "Hàng khuyến mãi"];
    if (tenGocUpper.includes("LÃI") && tenGocUpper.includes("VAY")) return ["FIN-AI", "Lãi ngân hàng/Lãi vay"];
    if (tenGocUpper.includes("THANH TOÁN")) return ["PAY-AI", "Thanh toán giao dịch"];
        
    return ["IT-GEN", "Phụ kiện / Linh kiện chung"];
}

const groups = {};

items.forEach(item => {
    if (item.sl < 2) return;
    const tenGoc = item.tenGoc.replace(/\n/g, ' ').trim();
    const [maHang, tenChuan] = getMapping(tenGoc);
    
    if (!groups[maHang]) {
        groups[maHang] = { ma: maHang, chuan: tenChuan, sl_tong: 0, gocs: new Set(), goc_list: [] };
    }
    
    const grp = groups[maHang];
    grp.sl_tong += item.sl;
    if (!grp.gocs.has(tenGoc)) {
        grp.gocs.add(tenGoc);
        if (grp.goc_list.length < 5) {
            grp.goc_list.push(tenGoc);
        }
    }
});

let sortedGroups = Object.values(groups);
sortedGroups.sort((a, b) => b.sl_tong - a.sl_tong);

let out = [];
out.push(`# Bảng Từ Điển Đồng Nghĩa (Synonyms) AI - CÔNG TY TNHH HUY VŨ`);
out.push("");
out.push(`Theo phân tích từ 1,968 mặt hàng phát sinh thực tế, AI đã gom nhóm và chuẩn hóa thành ${sortedGroups.length} danh mục chính. Bảng danh mục nhóm gọn nhẹ này giúp tối ưu cho việc cấu hình RAG Synonyms và tiết kiệm LLM tokens:`);
out.push("");
out.push("| STT | Mã Hàng (AI) | Tên Chuẩn Hóa | Số Lần Phát Sinh | Các Biến Thể Tên Hàng Thực Tế (Ví dụ) |");
out.push("| :--- | :--- | :--- | :--- | :--- |");

sortedGroups.forEach((grp, idx) => {
    let variants = "<br>- ".concat(grp.goc_list.join("<br>- "));
    if (grp.gocs.size > 5) {
        variants += `<br>*(...và ${grp.gocs.size - 5} biến thể khác)*`;
    }
    let s = `| ${idx + 1} | **${grp.ma}** | **${grp.chuan}** | ${grp.sl_tong} | ${variants.substring(4)} |`;
    out.push(s);
});

out.push("");
out.push(`*Ghi chú: Quá trình tinh gọn này đã giảm từ gần 2,000 danh mục thô xuống còn **${sortedGroups.length} mã chuẩn hóa chính**, giúp RAG hoạt động cực kỳ chính xác.*`);

fs.writeFileSync('/mnt/chikiet/kata2025/ragketoan/docs/huyvu/dulieuchuan/mapping_items.md', out.join('\n'));
console.log("Written successfully!");
